### ISSUE TITLE
Operator is not creating underlying k8s objects for Prometheus

### ISSUE THREAD

**Issue by rbs4ba** [prometheus-operator/prometheus-operator#3747]
Slack link for a bit of context: https://kubernetes.slack.com/archives/CFFDS2Z7F/p1608060016432200

**What happened?**
Recently we have noticed that it seems like the operator is not reflecting changes from the CRD objects in the k8s cluster. We first noticed that it seemed like new `serviceMonitors` weren't getting picked up. Upon doing a fresh install of the operator, now nothing is created from the `prometheus` object.

**Did you expect to see something different?**
I would expect that the Prometheus-Operator sees the `prometheus` custom-k8s-object and then creates the underlying k8s objects: the `statefulset, service, secret, and configmap`

**How to reproduce it (as minimally and precisely as possible)**:
Very tough to say. I can only imagine the issue is localized to me because I haven't seen any other issues posted about this. At the moment, we have reproduced this in a few of our clusters simply by doing a re-install of the operator via the helm chart. Because this affectively leaves us blind in said cluster (because there's no Prometheus application) we don't want to experiment too much in our other clusters that are still working.

**Environment**

* Prometheus Operator version:

    `Image:      docker.io/bitnami/prometheus-operator:0.44.1-debian-10-r0`

* Kubernetes version information:

    `Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.9-eks-d1db3c", GitCommit:"d1db3c46e55f95d6a7d3e5578689371318f95ff9", GitTreeState:"clean", BuildDate:"2020-10-20T22:18:07Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}`

* Kubernetes cluster kind:

    EKS

* Manifests:

Manifest for the `prometheus` object
```
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  labels:
    app.kubernetes.io/component: prometheus
    app.kubernetes.io/instance: kube-prometheus-in-cluster
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: kube-prometheus
    argocd.argoproj.io/instance: kube-prometheus-in-cluster
    helm.sh/chart: kube-prometheus-3.3.2
  name: kube-prometheus-in-cluster-prometheus
  namespace: monitoring
spec:
  affinity:
    podAntiAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
        - podAffinityTerm:
            labelSelector:
              matchLabels:
                app.kubernetes.io/component: prometheus
                app.kubernetes.io/instance: kube-prometheus-in-cluster
                app.kubernetes.io/name: kube-prometheus
            namespaces:
              - monitoring
            topologyKey: kubernetes.io/hostname
          weight: 1
  alerting:
    alertmanagers:
      - name: kube-prometheus-in-cluster-alertmanager
        namespace: monitoring
        pathPrefix: /
        port: http
  containers:
    - args:
        - sidecar
        - '--prometheus.url=http://localhost:9090'
        - '--grpc-address=0.0.0.0:10901'
        - '--http-address=0.0.0.0:10902'
        - '--tsdb.path=/prometheus/'
      image: 'docker.io/bitnami/thanos:0.17.2-scratch-r0'
      imagePullPolicy: IfNotPresent
      name: thanos-sidecar
      ports:
        - containerPort: 10901
          name: grpc
          protocol: TCP
        - containerPort: 10902
          name: http
          protocol: TCP
      resources:
        limits:
          cpu: 100m
          memory: 128Mi
        requests:
          cpu: 100m
          memory: 128Mi
      volumeMounts:
        - mountPath: /prometheus
          name: prometheus-kube-prometheus-in-cluster-prometheus-db
          subPath: prometheus-db
  enableAdminAPI: false
  evaluationInterval: 10s
  externalLabels:
    cluster: svc
    environment: dev-us-west-2
  externalUrl: '<url>'
  image: 'docker.io/bitnami/prometheus:2.23.0-debian-10-r6'
  listenLocal: false
  logFormat: logfmt
  logLevel: info
  nodeSelector:
    node-group: prometheus
  paused: false
  podMetadata:
    annotations:
      cluster-autoscaler.kubernetes.io/safe-to-evict: 'true'
    labels:
      app.kubernetes.io/component: prometheus
      app.kubernetes.io/instance: kube-prometheus-in-cluster
      app.kubernetes.io/name: kube-prometheus
  podMonitorNamespaceSelector: {}
  podMonitorSelector: {}
  probeNamespaceSelector: {}
  probeSelector: {}
  replicas: 1
  resources:
    limits:
      cpu: '4'
      memory: 20Gi
    requests:
      cpu: '4'
      memory: 20Gi
  retention: 7d
  routePrefix: /
  ruleNamespaceSelector: {}
  ruleSelector: {}
  scrapeInterval: 10s
  securityContext:
    fsGroup: 1001
    runAsUser: 1001
  serviceAccountName: kube-prometheus-in-cluster-prometheus
  serviceMonitorNamespaceSelector: {}
  serviceMonitorSelector: {}
  storage:
    volumeClaimTemplate:
      spec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 200Gi
        storageClassName: encrypted-gp2-csi
  walCompression: true
```

* Prometheus Operator Logs:
I can post full debug logs if desired. The only thing I've noticed in the debug logs is that 1 minute after the operator comes up, I see
```
level=warn ts=2020-12-16T17:28:19.594614676Z caller=operator.go:344 component=prometheusoperator informer=Secret msg="cache sync not yet completed"
```
and then about 30 seconds later I see
```
level=debug ts=2020-12-16T17:28:43.795861138Z caller=klog.go:60 component=k8s_client_runtime func=Info msg="Trace[469339106]: \"Reflector ListAndWatch\" name:pkg/mod/k8s.io/client-go@v0.19.2/tools/cache/reflector.go:156 (16-Dec-2020 17:27:17.696) (total time: 86099ms):\nTrace[469339106]: ---\"Objects listed\" 86097ms (17:28:00.794)\nTrace[469339106]: ---\"Resource version extracted\" 0ms (17:28:00.794)\nTrace[469339106]: ---\"Objects extracted\" 0ms (17:28:00.794)\nTrace[469339106]: ---\"SyncWith done\" 1ms (17:28:00.795)\nTrace[469339106]: ---\"Resource version updated\" 0ms (17:28:00.795)\nTrace[469339106]: [1m26.099660642s] [1m26.099660642s] END\n"
level=debug ts=2020-12-16T17:28:43.797611397Z caller=klog.go:72 component=k8s_client_runtime func=Infof msg="GET https://172.20.0.1:443/api/v1/secrets?allowWatchBookmarks=true&resourceVersion=77954356&timeout=6m0s&timeoutSeconds=360&watch=true 200 OK in 1 milliseconds"
level=debug ts=2020-12-16T17:28:43.894660254Z caller=klog.go:55 component=k8s_client_runtime func=Verbose.Infof msg="caches populated"
level=debug ts=2020-12-16T17:28:43.894726445Z caller=klog.go:72 component=k8s_client_runtime func=Infof msg="Caches are synced for prometheus "
level=debug ts=2020-12-16T17:28:43.894746326Z caller=operator.go:354 component=prometheusoperator informer=Secret msg="successfully synced cache"
```

Another thing I noticed in the logs, is that at the `info` level, there is no message `successfully synced all caches` for the `component=prometheusoperator` nor the `component=alertmanageroperator`, which does seem quite suspect.
```
level=info ts=2020-12-16T17:27:17.401538164Z caller=main.go:235 msg="Starting Prometheus Operator" version="(version=0.44.1, branch=, revision=)"
level=info ts=2020-12-16T17:27:17.401599605Z caller=main.go:236 build_context="(go=go1.15.6, user=, date=20201209-17:57:28)"
level=warn ts=2020-12-16T17:27:17.401611565Z caller=main.go:239 msg="'--config-reloader-image' flag is ignored, only '--prometheus-config-reloader' is used" config-reloader-image=docker.io/bitnami/configmap-reload:0.4.0-debian-10-r128 prometheus-config-reloader=docker.io/bitnami/prometheus-operator:0.44.1-debian-10-r0
ts=2020-12-16T17:27:17.595104445Z caller=main.go:107 msg="Starting insecure server on [::]:8080"
level=info ts=2020-12-16T17:27:17.695058197Z caller=operator.go:436 component=alertmanageroperator msg="connection established" cluster-version=v1.18.9-eks-d1db3c
level=info ts=2020-12-16T17:27:17.695093888Z caller=operator.go:445 component=alertmanageroperator msg="CRD API endpoints ready"
level=info ts=2020-12-16T17:27:17.695173549Z caller=operator.go:300 component=thanosoperator msg="connection established" cluster-version=v1.18.9-eks-d1db3c
level=info ts=2020-12-16T17:27:17.695187889Z caller=operator.go:309 component=thanosoperator msg="CRD API endpoints ready"
level=info ts=2020-12-16T17:27:17.695974882Z caller=operator.go:420 component=prometheusoperator msg="connection established" cluster-version=v1.18.9-eks-d1db3c
level=info ts=2020-12-16T17:27:17.695998713Z caller=operator.go:429 component=prometheusoperator msg="CRD API endpoints ready"
level=info ts=2020-12-16T17:27:19.593171184Z caller=operator.go:261 component=thanosoperator msg="successfully synced all caches"
level=warn ts=2020-12-16T17:28:19.594614676Z caller=operator.go:344 component=prometheusoperator informer=Secret msg="cache sync not yet completed"
level=warn ts=2020-12-16T17:28:19.793385527Z caller=operator.go:344 component=alertmanageroperator informer=Secret msg="cache sync not yet completed"
```

**Anything else we need to know?**:
Our current theory is that there is something wrong with the k8s `apiserver` itself. We have seen panics in the `apiserver` logs related to `metrics-server` scrapes, which we are investigating in parallel to this issue. So perhaps prometheus-operator needs to talk to the k8s `apiserver` to get the latest info on something related to the `prometheus` object, but it can't? Note, I have not seen anything in the `apiserver` logs that would seem to correlate to this issue with `prometheus-operator`

**Comment by rbs4ba** [comment]
An update, _some_ of our clusters are broken and some are okay, but we still don't know why.

In the working cluster, I was able to observe the operator logs. Notably I can see the log message for `successfully synced all caches`. I can also see in the debug logs that there are multiple cache syncs happening (`Prometheus, ServiceMonitor, PodMonitor, etc`). Compared to the broken clusters, I only see this sync for `Secrets` (these are the debug logs I originally posted above). So in the broken clusters, we're only able to sync the caches for `Secrets`; what does this mean? How is prometheus-operator doing this work? Is it talking to the master's `apiserver` or some other method?
```
level=info ts=2020-12-16T22:06:52.668795793Z caller=operator.go:420 component=prometheusoperator msg="connection established" cluster-version=v1.18.9-eks-d1db3c
level=info ts=2020-12-16T22:06:52.668822907Z caller=operator.go:429 component=prometheusoperator msg="CRD API endpoints ready"
level=debug ts=2020-12-16T22:06:53.956946969Z caller=operator.go:354 component=prometheusoperator informer=Prometheus msg="successfully synced cache"
level=debug ts=2020-12-16T22:06:53.957029429Z caller=operator.go:354 component=prometheusoperator informer=ServiceMonitor msg="successfully synced cache"
level=debug ts=2020-12-16T22:06:53.957103072Z caller=operator.go:354 component=prometheusoperator informer=PodMonitor msg="successfully synced cache"
level=debug ts=2020-12-16T22:06:53.957148444Z caller=operator.go:354 component=prometheusoperator informer=PrometheusRule msg="successfully synced cache"
level=debug ts=2020-12-16T22:06:53.957197164Z caller=operator.go:354 component=prometheusoperator informer=Probe msg="successfully synced cache"
level=debug ts=2020-12-16T22:06:53.957257351Z caller=operator.go:354 component=prometheusoperator informer=ConfigMap msg="successfully synced cache"
level=debug ts=2020-12-16T22:06:56.358116026Z caller=operator.go:354 component=prometheusoperator informer=Secret msg="successfully synced cache"
level=debug ts=2020-12-16T22:06:56.358155171Z caller=operator.go:354 component=prometheusoperator informer=StatefulSet msg="successfully synced cache"
level=debug ts=2020-12-16T22:06:56.358195811Z caller=operator.go:354 component=prometheusoperator informer=PromNamespace msg="successfully synced cache"
level=debug ts=2020-12-16T22:06:56.358222676Z caller=operator.go:354 component=prometheusoperator informer=MonNamespace msg="successfully synced cache"
level=info ts=2020-12-16T22:06:56.358228987Z caller=operator.go:359 component=prometheusoperator msg="successfully synced all caches"
```

**Comment by lilic** [comment]
Thanks for the bug report. Yes the Prometheus operator relies heavily on the apiserver to fetch all objects but also to create watches on them to get the updates, and afterward reconcile into the desired state.

> level=warn ts=2020-12-16T17:28:19.594614676Z caller=operator.go:344 component=prometheusoperator informer=Secret msg="cache sync not yet completed"

This does seem to confirm the theory around the apiserver. I do not believe we did anything to how we interact with the apiserver in the latest release. @prometheus-operator/prometheus-operator-reviewers if anyone knows more feel free to correct me!

Did you recently upgrade k8s version itself as well? Or just the version of the operator changed?

**Comment by rbs4ba** [comment]
We have solved this issue. In short, the solution was to give the prometheus-operator pod more CPU. After increasing the CPU we saw immediate improvements. At first we saw more of the log messages like
```
informer=Prometheus msg="successfully synced cache"
```
But, compared to a successful install, we did not see all of the caches being synced, nor did we see
```
component=prometheusoperator msg="successfully synced all caches"
```

Upon giving it slightly more CPU, everything started working again.

We first got the idea to limit the number of namespaces the operator was working with from the issue mentioned in the comment here: https://github.com/prometheus-operator/prometheus-operator/blob/master/pkg/operator/operator.go#L333. After limiting the namespaces, we also saw immediate improvements. At that point we figured the issue was related to how many objects were being handled in a given namespace. Which led us down the route noted above of giving the operator more resources.

Based on the logs, it seems like prometheus-operator was getting hung up on `secrets` in particular. We are surprised that one of the following didn't happen:
- 1) Additional logs that say something along the lines of "trying to get secret cache again"
- 2) The operator would eventually finish processing the secrets, even if it takes a long time
So we are curious what the operator might have been doing after the two following logs showed up
```
level=warn ts=2020-12-16T17:28:19.594614676Z caller=operator.go:344 component=prometheusoperator informer=Secret msg="cache sync not yet completed"
...
level=debug ts=2020-12-16T17:28:43.894746326Z caller=operator.go:354 component=prometheusoperator informer=Secret msg="successfully synced cache"
```

So all of this said, I imagine you might be able to recreate this situation if you give the operator pod a low amount of CPU and have many `secret` objects in the namespace(s) that the operator is handling.

Appreciate you taking the time to review all this! Would love to see how this sort of situation might be able to be handled more clearly, to let the user know that somehow the operator is stuck processing a cache (or whatever else it might be getting hung up on). Feel free to close this if you'd like. Happy to answer more questions.

**Comment by paulfantom** [comment]
Seems like this is not the first issue about performance problems around many secrets. Same was reported previously in https://github.com/prometheus-operator/prometheus-operator/issues/3298 and https://github.com/prometheus-operator/prometheus-operator/issues/3547. A recommendation is to use `--secret-field-selector` implemented in 0.42 in https://github.com/prometheus-operator/prometheus-operator/pull/3355. 

We should improve our [troubleshooting.md](https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/troubleshooting.md) document with those requirements.

**Comment by lilic** [comment]
Agreed to mention this in the troubleshooting! @rbs4ba would you be willing to contribute this to that document? 🎉

**Comment by simonpasquier** [comment]
> Would love to see how this sort of situation might be able to be handled more clearly, to let the user know that somehow the operator is stuck processing a cache (or whatever else it might be getting hung up on)

How could we improve the warning message stating `"cache sync not yet completed"`?

**Comment by madianas** [comment]
Had the same issue on two clusters. Error is not helpful at all. @rbs4ba  solution solved the problem.

**Comment by simonpasquier** [comment]
Closed by #4144 the operator should now log every minute a message stating that it's waiting on cache sync completion.
